### PR TITLE
algebra: adding additional tests for `matrix_prod` function

### PR DESCRIPTION
**Pull request by PNieck**
JIRA: PP-71

## Description

No tests checks cases with incorrect matrix sizes.

## Motivation and Context

More precise testing.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## How Has This Been Tested?

- [ ] Already covered by automatic testing.
- [ ] New test added: (add PR link here).
- [x] Tested by hand on: (list targets here). `armv7a9-zynq7000-qemu`

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing linter checks and tests passed.
- [ ] My changes generate no new compilation warnings for any of the targets.

## Special treatment

- [ ] This PR needs additional PRs to work (list the PRs, preferably in merge-order).
- [x] I will merge this PR by myself when appropriate.

### CODE REVIEW

File: algebra/tests/matrix/matrix_multiplication.c [hunk 1]
```diff
@@ -303,7 +313,106 @@ TEST(group_matrix_prod_bigMat, matrix_prod_bigMatsAllMatTrp)
...
+	TEST_ASSERT_EQUAL_INT(BUF_ALLOC_OK, algebraTests_matrixCopy(&M5, &M2));
+
+	TEST_ASSERT_EQUAL_INT(PRODUCT_OK, matrix_prod(&M1, &M2, &M3));
+
+	TEST_ASSERT_EQUAL_INT(CHECK_OK, algebraTest_equalMatrix(&M1, &M4));
+	TEST_ASSERT_EQUAL_INT(CHECK_OK, algebraTest_equalMatrix(&M2, &M5));
+}
+
+
+TEST_GROUP(group_matrix_prod_incorrMats);
```
Comment (niewim19) [inline algebra/tests/matrix/matrix_multiplication.c:329]: What do you think? Sounds more natural. May be substituted in all `incorr` occurrences
```suggestion
TEST_GROUP(group_matrix_prod_badMats);
```

File: algebra/tests/matrix/matrix_multiplication.c [hunk 2]
```diff
@@ -303,7 +313,106 @@ TEST(group_matrix_prod_bigMat, matrix_prod_bigMatsAllMatTrp)
...
+
+
+/* This test checks if result matrix is changing when function fails */
+TEST(group_matrix_prod_incorrMats, matrix_prod_failureRetain)
+{
+	M2.rows--;
+	M2.rows--;
+	TEST_ASSERT_EQUAL_INT(BUF_ALLOC_OK, matrix_bufAlloc(&M3, M1.rows, M2.cols));
+
+	TEST_ASSERT_EQUAL_INT(BUF_ALLOC_OK, algebraTests_matrixCopy(&M4, &M3));
```
Comment (niewim19) [inline algebra/tests/matrix/matrix_multiplication.c:410]: Gives better idea on what is the preparation part
```suggestion
	TEST_ASSERT_EQUAL_INT(BUF_ALLOC_OK, matrix_bufAlloc(&M3, M1.rows, M2.cols));
	TEST_ASSERT_EQUAL_INT(BUF_ALLOC_OK, algebraTests_matrixCopy(&M4, &M3));
```

File: algebra/tests/matrix/matrix_multiplication.c [hunk 3]
```diff
@@ -303,7 +313,106 @@ TEST(group_matrix_prod_bigMat, matrix_prod_bigMatsAllMatTrp)
...
+
+	TEST_ASSERT_EQUAL_INT(PRODUCT_FAIL, matrix_prod(&M1, &M2, &M3));
+}
+
+
+/* This test checks if result matrix is changing when function fails */
+TEST(group_matrix_prod_incorrMats, matrix_prod_failureRetain)
+{
+	M2.rows--;
+	M2.rows--;
```
Comment (niewim19) [inline algebra/tests/matrix/matrix_multiplication.c:407]: Why is this `-= 2` necessary? Shouldn`t it be;
```c
	M2.rows--;
	M2.cols--;
```

**Review by niewim19 — COMMENTED** [review]
Tests look very good. 

Apart from small naming change `incorr`->`bad` there is one thing i would like you to explain about double decrement of rows in M2 in one test.

File: algebra/tests/matrix/matrix_multiplication.c (hunk 3 above)
Comment (PNieck) [reply algebra/tests/matrix/matrix_multiplication.c:407]: Yes - that is my mistake. It should be
```c
	M2.rows--;
	M2.cols--;
```

File: algebra/tests/matrix/matrix_multiplication.c [hunk 4]
```diff
@@ -303,7 +313,105 @@ TEST(group_matrix_prod_bigMat, matrix_prod_bigMatsAllMatTrp)
...
+	matrix_bufFree(&M3);
+	matrix_bufFree(&M4);
+}
+
+
+TEST(group_matrix_prod_badMats, matrix_prod_badInputMats)
+{
+	M2.rows++;
+	M2.cols++;
+	TEST_ASSERT_EQUAL_INT(BUF_ALLOC_OK, matrix_bufAlloc(&M3, M1.rows, M2.cols));
```
Comment (niewim19) [inline algebra/tests/matrix/matrix_multiplication.c:355]: Isn't that a memory leak? You allocate buffer in `M3` without previousely freeing that buffer after allocation in `TEST_SETUP` so you `lost` that memory.

File: algebra/tests/matrix/matrix_multiplication.c [hunk 5]
```diff
@@ -303,7 +313,105 @@ TEST(group_matrix_prod_bigMat, matrix_prod_bigMatsAllMatTrp)
...
+	/* First and second transposed */
+	matrix_trp(&M1);
+
+	TEST_ASSERT_EQUAL_INT(PRODUCT_FAIL, matrix_prod(&M1, &M2, &M3));
+}
+
+
+TEST(group_matrix_prod_badMats, matrix_prod_badResMat)
+{
+	TEST_ASSERT_EQUAL_INT(BUF_ALLOC_OK, matrix_bufAlloc(&M3, M1.rows, M2.cols));
```
Comment (niewim19) [inline algebra/tests/matrix/matrix_multiplication.c:380]: Isn't that a memory leak? You allocate buffer in `M3` without previousely freeing that buffer after allocation in `TEST_SETUP` so you `lost` that memory.

File: algebra/tests/matrix/matrix_multiplication.c [hunk 6]
```diff
@@ -303,7 +313,105 @@ TEST(group_matrix_prod_bigMat, matrix_prod_bigMatsAllMatTrp)
...
+	TEST_ASSERT_EQUAL_INT(PRODUCT_FAIL, matrix_prod(&M1, &M2, &M3));
+}
+
+
+/* This test checks if result matrix is changing when function fails */
+TEST(group_matrix_prod_badMats, matrix_prod_failureRetain)
+{
+	M2.rows--;
+	M2.cols--;
+	TEST_ASSERT_EQUAL_INT(BUF_ALLOC_OK, matrix_bufAlloc(&M3, M1.rows, M2.cols));
```
Comment (niewim19) [inline algebra/tests/matrix/matrix_multiplication.c:408]: Isn't that a memory leak? You allocate buffer in `M3` without previousely freeing that buffer after allocation in `TEST_SETUP` so you `lost` that memory.

**Review by niewim19 — COMMENTED** [review]
I think that there is a memory leak in newly added tests. You are technically doing:
```c
matrix_bufAlloc(M3, ...); /* from TEST_SETUP */
/* (...) no free */
matrix_bufAlloc(M3, ...); /* from TEST_BODY */
```

**Comment by PNieck** [comment]
> I think that there is a memory leak in newly added tests. You are technically doing:
> 
> ```c
> matrix_bufAlloc(M3, ...); /* from TEST_SETUP */
> /* (...) no free */
> matrix_bufAlloc(M3, ...); /* from TEST_BODY */
> ```

Yes, you are right - my fault. I have repaired this

**Review by niewim19 — APPROVED** [review]
LGTM

**Pull request merged**